File: src/lib/catalog/skills.ts
/**
 * Skills (analytical playbooks). A skill is a markdown body the agent loads
 * into its system prompt when one of its triggers matches the user's question.
 * Triggers are simple keyword/phrase matches for v0.1; semantic matching
 * comes later.
 */
import { catalogPool } from "./db";

export type SkillRow = {
  id: number;
  slug: string;
  name: string;
  description: string;
  triggers: string[];
  body_md: string;
  enabled: boolean;
  created_at: string;
  updated_at: string;
};

export async function listSkills(): Promise<SkillRow[]> {
  const r = await catalogPool.query<SkillRow>(
    `SELECT id, slug, name, description, triggers, body_md, enabled, created_at, updated_at
       FROM skills ORDER BY name`,
  );
  return r.rows;
}

export async function getSkill(slug: string): Promise<SkillRow | null> {
  const r = await catalogPool.query<SkillRow>(
    `SELECT id, slug, name, description, triggers, body_md, enabled, created_at, updated_at
       FROM skills WHERE slug = $1`,
    [slug],
  );
  return r.rows[0] ?? null;
}

export async function upsertSkill(input: {
  slug: string;
  name: string;
  description: string;
  triggers: string[];
  body_md: string;
  enabled?: boolean;
}): Promise<SkillRow> {
  const r = await catalogPool.query<SkillRow>(
    `INSERT INTO skills (slug, name, description, triggers, body_md, enabled, updated_at)
       VALUES ($1, $2, $3, $4::jsonb, $5, COALESCE($6, TRUE), now())
     ON CONFLICT (slug) DO UPDATE
        SET name = EXCLUDED.name,
            description = EXCLUDED.description,
            triggers = EXCLUDED.triggers,
            body_md = EXCLUDED.body_md,
            enabled = COALESCE(EXCLUDED.enabled, skills.enabled),
            updated_at = now()
     RETURNING id, slug, name, description, triggers, body_md, enabled, created_at, updated_at`,
    [input.slug, input.name, input.description, JSON.stringify(input.triggers), input.body_md, input.enabled ?? null],
  );
  return r.rows[0]!;
}

export async function deleteSkill(slug: string): Promise<void> {
  await catalogPool.query(`DELETE FROM skills WHERE slug = $1`, [slug]);
}

/**
 * Find skills whose triggers appear (case-insensitive substring) in the
 * user's question. Returns a small set so we don't blow up the prompt.
 */
export async function matchSkills(question: string, limit = 3): Promise<SkillRow[]> {
  const skills = await listSkills();
  const lower = question.toLowerCase();
  const scored: { skill: SkillRow; hits: number }[] = [];
  for (const s of skills) {
    if (!s.enabled) continue;
    let hits = 0;
    for (const t of s.triggers) {
      if (typeof t === "string" && lower.includes(t.toLowerCase())) hits++;
    }
    if (hits > 0) scored.push({ skill: s, hits });
  }
  scored.sort((a, b) => b.hits - a.hits);
  return scored.slice(0, limit).map((s) => s.skill);
}

File: backend/app/catalog/skills.py
"""Skills (analytical playbooks).

Port of src/lib/catalog/skills.ts. A skill is a markdown body the agent loads
into its system prompt when one of its triggers matches the user's question.
"""
from __future__ import annotations

from typing import Any

from psycopg.rows import dict_row
from psycopg.types.json import Jsonb

from ..db import get_conn


async def list_skills() -> list[dict[str, Any]]:
    async with get_conn() as conn:
        async with conn.cursor(row_factory=dict_row) as cur:
            await cur.execute(
                """
                SELECT id, slug, name, description, triggers, body_md, enabled,
                       created_at, updated_at
                  FROM skills ORDER BY name
                """
            )
            return await cur.fetchall()


async def get_skill(slug: str) -> dict[str, Any] | None:
    async with get_conn() as conn:
        async with conn.cursor(row_factory=dict_row) as cur:
            await cur.execute(
                """
                SELECT id, slug, name, description, triggers, body_md, enabled,
                       created_at, updated_at
                  FROM skills WHERE slug = %s
                """,
                (slug,),
            )
            return await cur.fetchone()


async def upsert_skill(
    *,
    slug: str,
    name: str,
    description: str,
    triggers: list[str],
    body_md: str,
    enabled: bool | None = None,
) -> dict[str, Any]:
    async with get_conn() as conn:
        async with conn.cursor(row_factory=dict_row) as cur:
            await cur.execute(
                """
                INSERT INTO skills (slug, name, description, triggers, body_md, enabled, updated_at)
                     VALUES (%s, %s, %s, %s, %s, COALESCE(%s, TRUE), now())
                ON CONFLICT (slug) DO UPDATE
                  SET name = EXCLUDED.name,
                      description = EXCLUDED.description,
                      triggers = EXCLUDED.triggers,
                      body_md = EXCLUDED.body_md,
                      enabled = COALESCE(EXCLUDED.enabled, skills.enabled),
                      updated_at = now()
                RETURNING id, slug, name, description, triggers, body_md, enabled,
                          created_at, updated_at
                """,
                (slug, name, description, Jsonb(triggers), body_md, enabled),
            )
            row = await cur.fetchone()
    assert row is not None
    return row


async def delete_skill(slug: str) -> None:
    async with get_conn() as conn:
        async with conn.cursor() as cur:
            await cur.execute("DELETE FROM skills WHERE slug = %s", (slug,))


async def match_skills(question: str, limit: int = 3) -> list[dict[str, Any]]:
    """Score by simple case-insensitive trigger substring match."""
    skills = await list_skills()
    lower = question.lower()
    scored: list[tuple[int, dict[str, Any]]] = []
    for s in skills:
        if not s["enabled"]:
            continue
        hits = sum(
            1
            for t in (s.get("triggers") or [])
            if isinstance(t, str) and t.lower() in lower
        )
        if hits > 0:
            scored.append((hits, s))
    scored.sort(key=lambda x: x[0], reverse=True)
    return [s for _, s in scored[:limit]]
